### DISCUSSION TITLE
Doesn't rg support RegEx flags such as Ungreedy matching?

### DISCUSSION THREAD

**Discussion by StacktraceException** [BurntSushi/ripgrep#1698] [General] [answered]
It seems that a lazy positive lookahead is impossible, also it seems that there is no such (I guess redundant) thing, because the Ungreedy flag (make all matches lazy by default) accomplishes the task.

You can see for yourself with this PCRE implementation: regex101.com
```
rg -Pe '^\d\d?\d?\.\d\d?M.*(?=\.)'
```
for this input:

```
2.5M	./kopia-kolejna/recup_dir.17/b10482719.jpg
4.2M	./kopia-kolejna/recup_dir.18/b14523847.jpg
6.7M	./kopia-kolejna/recup_dir.18/b11322221.tif
```

matches upto the last dot.

Adding a question mark to the syntax:
```
rg -Pe '^\d\d?\d?\.\d\d?M.*(?=\.?)'
```
makes it match all the input.

The same happens on that implementation website. BUT enabling the U flag makes ```rg -Pe '^\d\d?\d?\.\d\d?M.*(?=\.?)'``` work   as wanted.

**Comment by BurntSushi** [comment]
Sorry I'm confused. What question are you asking here?

Non-greedy matching is unrelated and orthogonal to look around.

**Comment by StacktraceException** [comment]
Sorry but is this some Troll Post?
The examples show everything including how they are related to look-around.
What exactly are you confused about?

sob., 3 paź 2020 o 18:17 Andrew Gallant <notifications@github.com>
napisał(a):

**Reply by BurntSushi** [reply]
If you accuse me of trolling you on my own issue tracker again, then you won't be welcome here. I'm trying to help. I simply do not understand what question you're asking or what problem you're trying to solve.

**Comment by BurntSushi** [comment, marked as answer]
Maybe all you're asking is how to use non-greedy matching. If so, I'll answer that, but your post is really confusing.

If you want to use non-greedy matching, then you have two options. The first is to add a question mark to the repetition operator that you want to be non-greedy. e.g., `??`, `+?` and `*?` are the non-greedy versions of `?`, `+` and `*`. You can also use the `U` flag to invert the greediness of *all* repetition operators.

As far as I know, this is standard across all regex engines that support greediness.

**Reply by StacktraceException** [reply]
Hi, I gave up on using ripgrep, never found a way (and I searched lots! on google and in your docs) to use the PCRE2 U flag. It seems to be nonexistent in ripgrep. In ripgrep's documentation you have ripgrep -U configured to behave as PCRE2's M (multiline) flag. 

I migrated to using pcre2regex from pcre2libs.



I'd still very appreciate it though if the confusion could be finally lifted??

**Reply by BurntSushi** [reply]
I would be happy to fix you confusion but I don't understand the problem. I think it would help if you showed the commands you're running, their output and the output you want.

The -P/--pcre2 flag enables PCRE2 in ripgrep.

**Reply by StacktraceException** [reply]
I'm happy that you'd be happy to do this. 

--------------------What commands I run and their output--------------------
(This command below originates from the explanation in my first post)

cat /rg-test.org|rg -Pe '^\d\d?\d?\.\d\d?M.*(?=\.)'

The input text is

2.5M	./kopia-kolejna/recup_dir.17/b10482719.jpg
4.2M	./kopia-kolejna/recup_dir.18/b14523847.jpg
6.7M	./kopia-kolejna/recup_dir.18/b11322221.tif

The match of this command is everything up to (but not including the last dot [which means the entire line is matched except the file extension]).

For the command cat /rg-test.org|rg -Pe '^\d\d?\d?\.\d\d?M.*(?=\.?)' all of the input text is matched.

--------- What is the expectation-----------

I need to make the command cat /rg-test.org|rg -Pe '^\d\d?\d?\.\d\d?M.*(?=\.?)' have Ungreedy matching so it will match only:

2.5M
4.2M
6.7M

(I'm not sure if this was what exactly I tried to accomplish, would be too time consuming to remember if even possible and/or useful, but anyway all of this shows the difference that the PCRE2's Ungreedy flag makes.

I tried several things to enable this PCRE2's Ungreedy flag but it seems that it in ripgrep it is not supported. It is not mentioned anywhere in the documentation, nor is there anything useful about this in google, nor did all my attempts to guess a correct syntax to use it ever work.

**Reply by BurntSushi** [reply]
Why not use `rg -o '^\d\d?\d?.\d\d?M'`. Neither PCRE2 nor ungreediness seem relevant here?

And ripgrep certainly does nothing to disable ungreediness. It works in both the default and PCRE2 regex engines:

```
$ echo foo | rg -o '\w+?'
1:f
$ echo foo | rg -Po '\w+?'
1:f
```

**Reply by StacktraceException** [reply]
Greediness is the default. So ripgrep also does nothing to enable it. The example you've just wrote omits what the commands and outputs you've asked for show. Maybe try to follow through what I wrote by running these commands with that input yourself. Nevertheless a potentially easier to notice difference may be that (as you can find in the just aforementioned now examples) 

.* matches as many characters as it can unless the greedy option is turned on. Regardless whether the example (like I've just wrote) really needs PCRE2's Ungreedy flag (or whether the expected match could have been matched without PCRE2's Ungreedy flag) I really did need it when I asked the question. And surely it is a useful feature, so please finally tell me how to use it in ripgrep.

**Reply by StacktraceException** [reply]
Sorry if I seem rude, I'm actually trying to help us have a better understanding

**Reply by BurntSushi** [reply]
`.*?` makes it ungreedy.

And yes, you are coming across as rude.

Not sure what else to say here, sorry. I do not understand your confusion.

**Reply by StacktraceException** [reply]
Ok, so today's example I've given was good only for explaining the PCRE2's Ungreedy flag, but it didn't prove its need. Sorry about this. Sorry also for coming off as rude, I'm not meaning to be.
Regardless if you see notice a need for PCRE2's Ungreedy flag (which is different than just adding a question mark to the syntax) please just tell how to use the  PCRE2's Ungreedy flag in ripgrep

**Reply by BurntSushi** [reply]
> PCRE2's Ungreedy flag (which is different than just adding a question mark to the syntax)

How is it different? Could you show an example, perhaps, using pcre2regex? (Or did you mean to reference pcre2grep before?)

**Reply by okdana** [reply]
> please just tell how to use the PCRE2's Ungreedy flag in ripgrep

Are you talking about `(?U)`? As in, `(?U)\w+` is equivalent to `\w+?`?

**Reply by StacktraceException** [reply]
Sorry, I'm giving up on explaining anymore. Please just tell me how to use the PCRE2's Ungreedy flag in ripgrep.

**Reply by BurntSushi** [reply]
I legitimately don't know what you're talking about if the "ungreedy flag" isn't the same as `.*?`. The question mark makes the repetition ungreedy. That is its definition. As @okdana said, `(?U:.*)` is equivalent to `.*?`.

I mean, if I understood your question I would answer it. I'm not trying to be coy here.

**Reply by StacktraceException** [reply]
> > please just tell how to use the PCRE2's Ungreedy flag in ripgrep
> 
> Are you talking about `(?U)`? As in, `(?U)\w+` is equivalent to `\w+?`?

This is most probably what I had in mind. (it almost certainly is, I'd just need to test it out with what I have had tried to do when asking this question). Thank You.
